### PR TITLE
[wasm] Outline exception throws in the interp and ifdef out unreachable opcodes

### PR DESCRIPTION
**Pull request by kg**
In my testing it improves performance if we outline exception throwing logic and don't include cases for unreachable opcodes. Based on a dump of the resulting dotnet.wasm module, this makes interp_exec_method ~3kb smaller and reduces the nesting depth by a bit as well.

According to my local browser-bench runs this makes the interpreter much faster for some workloads - I saw a consistent 1ms speed-up on some of the JSON test cases. My assumption is that interp_exec_method is big enough and complex enough that we are falling off performance cliffs and this somehow gets us back up one of the cliffs.

For unreachable opcodes the solution used here is to define a new `IROPDEF` category for opcodes that are meant for internal use and not meant to be executed - they all get moved to the end of the list and don't need cases defined or entries in the computed goto table, which makes interp_exec_method a bit smaller as well.

There are other opcodes that would benefit from this but I found that changing them caused weird test failures, so we can do that later.

### CODE REVIEW

File: src/mono/mono/mini/interp/interp.c [hunk 1]
```diff
@@ -3794,19 +3806,28 @@ mono_interp_exec_method (InterpFrame *frame, ThreadContext *context, FrameClause
...
 			MINT_IN_BREAK;
+#if HOST_BROWSER
+#else
 		MINT_IN_CASE(MINT_NOP)
 		MINT_IN_CASE(MINT_IL_SEQ_POINT)
 		MINT_IN_CASE(MINT_NIY)
 		MINT_IN_CASE(MINT_DEF)
 		MINT_IN_CASE(MINT_DUMMY_USE)
 		MINT_IN_CASE(MINT_TIER_PATCHPOINT_DATA)
+		MINT_IN_CASE(MINT_CALLRUN)
```
Comment (BrzVlad) [inline src/mono/mono/mini/interp/interp.c:3817]: MINT_CALLRUN should be deleted altogether, it no longer exists.

File: src/mono/mono/mini/interp/interp.c [hunk 2]
```diff
@@ -3794,19 +3806,28 @@ mono_interp_exec_method (InterpFrame *frame, ThreadContext *context, FrameClause
 			memset (locals + ip [1], 0, ip [2]);
 			ip += 3;
 			MINT_IN_BREAK;
+#if HOST_BROWSER
```
Comment (BrzVlad) [inline src/mono/mono/mini/interp/interp.c:3809]: I think we should have this also on desktop. The problem is that on desktop we use computed goto so labels for all opcodes must be declared. A possible approach would be to start separating between IR level opcodes, like these, and normal interp opcodes. These IR level opcodes should be allocated after all normal opcodes so they don't waste space in the opcode table. We could define them separately with something like `OPDEF_IR`, or in a separate file, or something else if you have any ideas.

File: src/mono/mono/mini/interp/mintops.def [hunk 3]
```diff
@@ -6,19 +6,39 @@
...
-OPDEF(MINT_DUMMY_USE, "dummy_use", 2, 0, 1, MintOpNoArgs)
-OPDEF(MINT_TIER_PATCHPOINT_DATA, "tier_patchpoint_data", 2, 0, 0, MintOpShortInt)
+// Compatibility shims for old code
+
+#ifndef IROPDEF
+#define __DEFINED_IROPDEF__
+#define IROPDEF(opsymbol, opstring, oplength, num_dregs, num_sregs, optype) OPDEF(opsymbol, opstring, oplength, num_dregs, num_sregs, optype)
+#endif // IROPDEF
+
+#ifndef WASMOPDEF
```
Comment (BrzVlad) [inline src/mono/mono/mini/interp/mintops.def:27]: I'm not sure I understand the utility of this def. Couldn't we just guard opcodes with normal platform ifdefs both in `mintops.def` and in `interp_exec_method` ? Seems more straightforward and maybe at some point we will have platform specific opcodes for other arches also.

Comment (kg) [reply src/mono/mono/mini/interp/mintops.def:27]: We can do that, yeah. It requires genmintops to understand ifdef though potentially

**Review by BrzVlad — APPROVED** [review]

**Pull request merged**